contracts/SwapAndBridge.sol

```diff
@@ -0,0 +1,150 @@
+     *     - srcReceiver: Address to receive `srcToken` from msg.sender. Overwritten by this function to be this contract.
+     *     - dstReceiver: Address to receive `dstToken` from 1inch swap. Overwritten by this function to be this contract.
+     *     - amount: Amount of `srcToken` to swap. Return amount is deposited into Across.
+     *     - minReturnAmount: Minimum amount of `dstToken` to receive from 1inch swap.
+     *     - flags: Overwritten by this contract.
+     * @param outputToken Token to receive from Across deposit on destination chain.
+     * @param outputAmount of `outputToken` to receive from Across deposit on
+     * destination chain.
+     */
+    function swap1InchAndBridge(
+        IAggregationExecutor aggregationExecutor,
+        I1InchAggregationRouterV5.SwapDescription memory swapDescription,
+        address outputToken,
+        uint256 outputAmount,
+        address depositor,
+        address recipient,
+        uint256 destinationChainid,
+        address exclusiveRelayer,
+        uint32 quoteTimestamp,
+        uint32 fillDeadline,
+        uint32 exclusivityDeadline,
+        bytes memory message
+    ) external nonReentrant {
+        // @dev: Don't let caller use WETH, or better, hardcode the swap and acrossInput tokens to be USDC.e/USDC
+        // because 1Inch has special rules to handle WETH<-->ETH.
+
+        // Pull tokens from caller into this contract.
+        address swapToken = address(swapDescription.srcToken);
+        uint256 swapTokenAmount = swapDescription.amount;
+        IERC20(swapToken).transferFrom(msg.sender, address(this), swapTokenAmount);
+
+        // Craft 1Inch swap arguments to swap `swapToken` for `acrossInputToken` using this contract as the recipient
+        // of the swap.
+        swapDescription.srcReceiver = payable(address(this));
+        swapDescription.dstReceiver = payable(address(this));
+        swapDescription.flags = 0; // TODO: Change this if we want to support ETH swaps and/or partial fills (i.e.
+        // swapTokenAmount != spentAmount). For now, set to 0 for simplicity.
+
+        IERC20(swapDescription.dstToken).approve(address(oneInchRouter), swapTokenAmount);
+
+        // @dev: Example swap I used for comparison:
+        // https://optimistic.etherscan.io/tx/0x8a4e77ee1a62e94b42b21e849bcdabd60d43ac49191cd2878f6b47f395f26abc
+        (uint256 returnAmount, uint256 spentAmount) = oneInchRouter.swap(
+            aggregationExecutor,
+            swapDescription,
+            new bytes(0), // TODO: No IERC20Permit.permit needed since we're sending an approval?
+            new bytes(0) // TODO: We don't want to execute any data on swaps but I'm not sure how this is used.
+        );
+
+        // TODO: I don't think we need the following check if we set flags to 0 but just in case:
+        require(spentAmount == swapTokenAmount, "spent amount != swap amount");
```

Comment:
There's three checks we could do here:
1. spentAmount == swapTokenAmount, which sort of checks that they follow their own rule
2. balanceIncrease >= minExpectedReturn (maybe expectedReturn is passed in separately?)
3. balanceDecrease == swapTokenAmount

I think 2 is probably the only one we _really_ need, since all the user cares about is that at least their min tokens are sent.

Thoughts?

Reply:
+1 on using point 2 that's a really good check. I don't think check 3 matters as much if we pass check 2 and check 1: however adding it does protect against some weird malicious case where 1inch returns an invalid `spentAmount`, which ultimately leads us to locking swapTokens in the contract forever.

Reply:
d5ba1dae353361f71de766a2acb63cae8b7bd2c9